src/code/sched.c

```diff
@@ -320,36 +445,45 @@ void Sched_HandleRetrace(SchedContext* sc) {
         osSyncPrintf("%08d:scHandleRetrace %08x\n", (u32)OS_CYCLES_TO_USEC(osGetTime()), osViGetCurrentFramebuffer());
     }
     ViConfig_UpdateBlack();
-    sc->retraceCnt++;
+    sc->retraceCount++;
 
-    if (osViGetCurrentFramebuffer() == (u32*)(sc->pendingSwapBuf1 != NULL ? sc->pendingSwapBuf1->swapBuffer : NULL)) {
+    // Retrace handlers run after VI context swap. The last swap buffer may now be the current buffer.
+    if (osViGetCurrentFramebuffer() == ((sc->pendingSwapBuf1 != NULL) ? sc->pendingSwapBuf1->swapBuffer : NULL)) {
         if (sc->curBuf != NULL) {
             sc->curBuf->unk_10 = 0;
         }
-
         if (sc->pendingSwapBuf1 != NULL) {
             sc->pendingSwapBuf1->unk_10 = 0;
         }
+        // Advance buffers
         sc->curBuf = sc->pendingSwapBuf1;
         sc->pendingSwapBuf1 = sc->pendingSwapBuf2;
         sc->pendingSwapBuf2 = NULL;
     }
+
     if (sc->curBuf != NULL) {
-        if (sc->curBuf->updateRate2 > 0) {
-            sc->curBuf->updateRate2--;
+        // Swap the framebuffer when the update timer runs out
+        if (sc->curBuf->updateTimer > 0) {
+            sc->curBuf->updateTimer--;
         }
-        if ((sc->curBuf->updateRate2 <= 0) && (sc->pendingSwapBuf1 != NULL)) {
-            func_800C84E4(sc, sc->pendingSwapBuf1);
+        if (sc->curBuf->updateTimer <= 0 && sc->pendingSwapBuf1 != NULL) {
+            Sched_RetraceSwapBuffer(sc, sc->pendingSwapBuf1);
         }
     }
+
     if (sLogScheduler) {
         osSyncPrintf("%08x %08x %08x %d\n", osViGetCurrentFramebuffer(), osViGetNextFramebuffer(),
-                     sc->pendingSwapBuf1 != NULL ? sc->pendingSwapBuf1->swapBuffer : NULL,
-                     sc->curBuf != NULL ? sc->curBuf->updateRate2 : 0);
+                     (sc->pendingSwapBuf1 != NULL) ? sc->pendingSwapBuf1->swapBuffer : NULL,
+                     (sc->curBuf != NULL) ? sc->curBuf->updateTimer : 0);
     }
-    Sched_HandleEntry(sc);
+
+    // Run the notification handler to enqueue any waiting tasks and possibly run one
+    Sched_HandleNotification(sc);
```

Comment:
it looks like `Sched_HandleNotification` could use a more generic name like `Sched_QueueAndRunTasks` / `Sched_HandleTasks` ?

Reply:
I've been thinking about it from the perspective of the retrace event itself also being a "notification", but I'll think on it

Reply:
Since `HandleNotification` ends in the same `Sched_Schedule`-using code as other `HandleXxx` functions it's probably fine as `HandleNotification`